### PR TITLE
feat(parser): distinguish exponential number

### PR DESCRIPTION
**Pull request by mysteryven**
We treat all exponential number(e.g. `1e1`) as Float now. In this PR, we distinguish it to Float or Decimal.

### CODE REVIEW

File: crates/oxc_parser/src/lexer/mod.rs [hunk 1]
```diff
@@ -841,7 +843,11 @@ impl<'a> Lexer<'a> {
             builder.push_matching(c);
         }
         self.read_decimal_digits(builder);
-        Kind::Float
+        if c == '-' {
```
Comment (anonrig) [inline crates/oxc_parser/src/lexer/mod.rs:846]: `~0` is `-1`. Using `-` delimiter is not sufficient to distinguish negative numbers.

File: crates/oxc_parser/src/lexer/kind.rs [hunk 2]
```diff
@@ -169,6 +169,8 @@ pub enum Kind {
     Binary,
     Octal,
     Hex,
+    PositiveExponential,
```
Comment (Boshen) [inline crates/oxc_parser/src/lexer/kind.rs:172]: Can you add a comment: `// for `1e10`, `1e+10`, `// for `1e-10`

File: crates/oxc_parser/src/lexer/mod.rs [hunk 3]
```diff
@@ -841,7 +843,7 @@ impl<'a> Lexer<'a> {
             builder.push_matching(c);
         }
         self.read_decimal_digits(builder);
-        Kind::Float
+        if c == '-' { Kind::NegativeExponential } else { Kind::PositiveExponential }
```
Comment (Boshen) [inline crates/oxc_parser/src/lexer/mod.rs:846]: Nit: make the code cleaner (and less if checks) by combining this check with the `matches` above.

**Review by Boshen — CHANGES_REQUESTED** [review]

File: crates/oxc_parser/src/lexer/kind.rs [hunk 4]
```diff
@@ -604,6 +609,8 @@ impl Kind {
             Binary => "binary",
             Octal => "octal",
             Hex => "hex",
+            PositiveExponential => "positiveexponential",
+            NegativeExponential => "negativeexponential",
```
Comment (Boshen) [inline crates/oxc_parser/src/lexer/kind.rs:613]: This is for error diagnostics, so both of them are still called `float`

**Pull request merged**